### ISSUE TITLE
ImageCanvas source canvasFunction not called after change on RasterSource

### ISSUE THREAD

**Issue by EvertEt** [openlayers/openlayers#11176]
**Describe the bug**
We notice sometimes the canvasFunction of an ol/source/ImageCanvas is not called again when calling `.changed` on the parent Raster source.
**To Reproduce**
1.  Log in the canvas function
```js
const canvasFunction = (extent, resolution, pixelRatio, size) => {
console.log("canvasFunction");
...
const canvasSource = new OlImageCanvasSource({ projection, canvasFunction });
canvasSource.on("change", () => console.log("canvasSource change"));
...
```
2. Construct Rastersource from ImageCanvas
```js
const rasterSource = new OlRasterSource({
  sources: [canvasSource],
   ...rasterOptions,
});
```
3. 
```js
new OlImageLayer({
  source: rasterSource,
});
```
4. 
```js
this.image.onload = () => {
  console.log("onload");
  // Refresh sources when image is loaded
  rasterSource.changed();
};
```

5. It seems like a race condition, normal logs with image showing up:
```
onload
canvasSource change
canvasFunction
```

But sometimes the map stays blank until you move it or zoom and it has these logs
```
onload
canvasSource change
```

**Expected behavior**
No race condition causing canvasFunction not to be called although changed is called.

**Comment by EvertEt** [comment]
I cannot reproduce it when putting some delay (1ms does not work, 100 seems to work)
```js
        this.image.onload = () => {
            // Refresh sources when image is loaded
            setTimeout(() => { // https://github.com/openlayers/openlayers/issues/11176
                canvasSource.changed();
            }, 100);
        };
```

**Comment by EvertEt** [comment]
Some further debugging which is hopefully helpful:
- `getImageInternal` does not get called as well in the broken case.
- ol/source/Image `getImage` doesn't get called as well.
- afteroperations event does get called on the rasterSource.
- Main cause seems to be that here, the revisions seem to match: https://github.com/openlayers/openlayers/blob/323a472fbbcc10c3a5040f3a1a449b4dbce331ba/src/ol/source/Raster.js#L699

**Comment by EvertEt** [comment]
I think my suspicion is indeed the case:
```
        this.worker_.process(imageDatas, data, this.onWorkerComplete_.bind(this, frameState, this.getRevision()));
```
Passes the revision when starting the worker.

```
    RasterSource.prototype.onWorkerComplete_ = function (frameState, revision, err, output, data) {
```

```
        this.renderedRevision_ = revision; // THIS LINE
        console.log("started at", revision, "to", this.renderedRevision_)
```
(these code snippets are from the compiled version which I'm editing to test live)

I believe changing `THIS LINE` to set the revision for which the worker was started instead of the revision at which we are when the worker completes can solve this. The question is if we still want to do this update if we're behind anyway?

**Comment by EvertEt** [comment]
I took a different approach to solve this.

Summary of bug:
1. Create OlImageCanvasSource with canvasFunction
2. Load image async for use in canvasFunction
3. First worker call starts (W1)
4. Image loads and I called `changed`
5. Second worker call starts (W2)
6. It seems that second result doesn't get rendered anymore, so the image only shows up after another re-render (zoom etc).

How I solved it now is by putting the `OlImageCanvasSource` in `OlSourceState.LOADING` before the image is loaded so W1 is skipped. I guess this bug can still occur but at least it removes it in my case (fast async loading after constructing) so not sure if I should close this?
